compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,186 +75,490 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
-    // In a loop, using the semi-affine expression:
-    // `iv % size + offset`,
-    // `iv` is the loop induction variable,
-    // `size` is the number of BD IDs assigned to each DMA op,
-    // `offset` is the BD ID assigned to current DMA op at first iteration.
-    Value iv = loop.getInductionVar();
-
-    // Get the number of BD IDs will be assigned to current DMA op.
-    uint32_t numRequired = 0;
-    getNumRequiredBdIds(loop, npuDmaOp, tileOp, shimTileToGeneratorMap,
-                        numRequired);
-    uint32_t numAvailable = generator.getNumAvailableBdIds(channel);
-    uint32_t size = std::max(numAvailable / numRequired, 1u);
-
-    // Only create expression if more than 1 BD ID is needed,
-    // otherwise, fall back to constant BD ID.
-    if (size > 1) {
-      // Assigning BD IDs for all iterations in the loop.
-      SmallVector<uint32_t> bdIds;
-      for (uint32_t i = 0; i < size; i++) {
-        std::optional<uint32_t> bdId = generator.getAndAssignBdId(
-            channel, BdIdAssignmentMode::Incremental);
-        if (!bdId) return failure();
-        bdIds.push_back(bdId.value());
-      }
-      // Get the BD ID for the first iteration as the offset.
-      uint32_t offset = bdIds.front();
 
-      // Create the semi-affine expression.
-      AffineExpr ivExpr;
+  /// Since a DmaBatch contains the following :-
+  ///   1. DmaOps within the current batch.
+  ///   2. Nested DmaBatch chain.
+  ///   3. Pointer to the sibling DmaBatch.
+  /// This API processes a given DmaBatch by :-
+  ///   a. Assigning Bd Ids to 1.
+  ///   b. Assigning Bd Ids to 2.
+  ///   c. Releasing Bd Ids assigned to 1.
+  ///   d. Moving on to 3 and repeating steps a-to-d for it.
+  LogicalResult processDmaBatch(IRRewriter &rewriter, AMDAIE::TileOp tileOp,
+                                std::shared_ptr<DmaBatch> currDmaBatch) {
+    do {
+      if (currDmaBatch->isEmpty()) break;
+      if (failed(assignRequiredBdIdsInCurrentBatch(rewriter, tileOp,
+                                                   currDmaBatch)))
+        return failure();
+
+      if (failed(
+              assignRequiredBdIdsInInnerBatch(rewriter, tileOp, currDmaBatch)))
+        return failure();
+
+      if (failed(
+              releaseAssignedBdIdsInCurrentBatch(currDmaBatch->currentDmaOps)))
+        return failure();
+
+      currDmaBatch = currDmaBatch->nextDmaBatch;
+    } while (currDmaBatch != nullptr);
+    return success();
+  }
+
+  /// Assign Bd Ids to the inner DmaBatch of the current DmaBatch
+  /// `parentDmaBatch`. And maintain a mapping of DmaOp -> source/target Bd Id
+  /// which would be used later during new DmaOp creation/replacement.
+  LogicalResult assignRequiredBdIdsInInnerBatch(
+      IRRewriter &rewriter, AMDAIE::TileOp tileOp,
+      std::shared_ptr<DmaBatch> parentDmaBatch) {
+    for (std::shared_ptr<DmaBatch> &dmaBatch :
+         parentDmaBatch->immediateInnerBatches) {
+      OpBuilder::InsertionGuard guard(rewriter);
```

Comment:
It's better to put the insertion guard in `processBatch` itself, then it would work in all cases in which the function is called.